lib/active_record/connection_adapters/cockroachdb/transaction_manager.rb

```diff
@@ -10,21 +10,44 @@ module TransactionManagerMonkeyPatch
         # exceeded.
         def within_new_transaction(isolation: nil, joinable: true, attempts: 0)
           super(isolation: isolation, joinable: joinable)
+        rescue ActiveRecord::ConnectionNotEstablished => error
+          raise unless retryable? error
+          raise if attempts >= @connection.max_transaction_retries
+
+          sleep_rand_seconds()
+
+          unless @connection.active?
+            warn "connection isn't active, reconnecting"
+            @connection.reconnect!
+          end
+
+          within_new_transaction(isolation: isolation, joinable: joinable, attempts: attempts + 1) { yield }
         rescue ActiveRecord::StatementInvalid => error
           raise unless retryable? error
           raise if attempts >= @connection.max_transaction_retries
 
-          attempts += 1
-          sleep_seconds = (2 ** attempts + rand) / 10
-          sleep(sleep_seconds)
-          within_new_transaction(isolation: isolation, joinable: joinable, attempts: attempts) { yield }
+          sleep_rand_seconds()
+
+          within_new_transaction(isolation: isolation, joinable: joinable, attempts: attempts + 1) { yield }
         end
 
         def retryable?(error)
+          return true if serialization_error?(error)
           return true if error.is_a? ActiveRecord::SerializationFailure
           return retryable? error.cause if error.cause
           false
         end
+
+        def serialization_error?(error)
+          errors = [error]
+          errors << error.cause if error.cause
+          errors.any? {|e| e.is_a? PG::TRSerializationFailure }
+        end
+
+        def sleep_rand_seconds()
+          sleep_seconds = (2 ** attempts + rand) / 10
```

Comment:
`attempts` should be a parameter of the method.

Reply:
Fixed.